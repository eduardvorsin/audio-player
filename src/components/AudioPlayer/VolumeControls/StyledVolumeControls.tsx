import styled from "styled-components";
import { StyledButton } from "../../UI/Button/StyledButton";

interface StyledVolumeControlsProps {
	className?: string;
}

export const StyledVolumeControls = styled.div<StyledVolumeControlsProps>`
	display: flex;
	align-self: flex-end;
	align-items: center;

	& > ${StyledButton}{
		flex: 0 0 40px;
		margin-right:10px;
	}
`;

