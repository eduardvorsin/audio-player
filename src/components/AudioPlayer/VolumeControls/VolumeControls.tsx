import React from "react";
import { Button } from "../../UI/Button/Button";
import { StyledVolumeControls } from "./StyledVolumeControls";
import { ReactComponent as MuteIcon } from '../../../assets/images/icons/mute.svg';
import { ReactComponent as VolumeIcon } from '../../../assets/images/icons/volume.svg';
import { RangeSlider } from "../../UI/RangeSlider/RangeSlider";

interface VolumeControlsProps {
	isMuted: boolean,
	volume?: number,
	onVolumeChange?: React.ChangeEventHandler<HTMLInputElement>,
	toggleMuting?: React.MouseEventHandler<HTMLButtonElement>,
}

export const VolumeControls = React.memo<VolumeControlsProps>(({
	volume = 50,
	onVolumeChange = () => { },
	isMuted = false,
	toggleMuting = () => { },
}) => {
	const currentIcon = isMuted ? <MuteIcon /> : <VolumeIcon />;

	return (
		<StyledVolumeControls
			data-testid='volume-controls'
		>
			<Button
				withoutVisibleText
				startIcon={currentIcon}
				onClick={toggleMuting}
			>
				{isMuted ? 'unmute' : 'mute'}
			</Button>
			<RangeSlider
				onChange={onVolumeChange}
				defaultValue={volume}
			/>
		</StyledVolumeControls>
	);
});
