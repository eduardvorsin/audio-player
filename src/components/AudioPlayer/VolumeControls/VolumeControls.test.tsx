import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { VolumeControls } from "./VolumeControls";

describe('Volume Controls component tests', () => {
  test('renders correctly', () => {
    const mockClickHandler = jest.fn<
      void, [React.MouseEvent<HTMLButtonElement>]
    >();
    const mockChangeHandler = jest.fn<
      void, [React.ChangeEvent<HTMLInputElement>]
    >();
    render(
      <VolumeControls
        isMuted
        volume={30}
        onVolumeChange={mockChangeHandler}
        toggleMuting={mockClickHandler}
      />
    );

    expect(screen.getByRole<HTMLButtonElement>('button')).toBeInTheDocument();
    expect(screen.getByRole<HTMLInputElement>('slider')).toBeInTheDocument();
  });

  test('renders with isMuted prop', () => {
    render(
      <VolumeControls
        isMuted
      />
    );

    expect(screen.getByRole<HTMLButtonElement>('button')).toHaveTextContent('mute');
    expect(screen.getByRole<HTMLInputElement>('slider')).toHaveValue('50');
  });

  test('when the button is clicked, the mock function is triggered', async () => {
    const user = userEvent.setup();
    const mockClickHandler = jest.fn<
      void, [React.MouseEvent<HTMLButtonElement>]
    >();
    render(
      <VolumeControls
        isMuted
        toggleMuting={mockClickHandler}
      />
    );

    await user.click(screen.getByRole<HTMLButtonElement>('button'));

    expect(mockClickHandler).toBeCalledTimes(1);
  });

  test('when the value of the range slider changed the mock function is triggered', () => {
    const mockChangeHandler = jest.fn<
      void, [React.ChangeEvent<HTMLInputElement>]
    >();
    render(
      <VolumeControls
        isMuted
        onVolumeChange={mockChangeHandler} />
    );

    fireEvent.change(screen.getByRole<HTMLInputElement>('slider'), { target: { value: 51 } });

    expect(mockChangeHandler).toBeCalledTimes(1);
  });

  test('when the isMuted prop is passed, the text in the button changes to unmute', () => {
    render(
      <VolumeControls isMuted />
    );

    expect(screen.getByRole<HTMLButtonElement>('button')).toHaveTextContent(/unmute/i);
  });

  test('snapshot with default prop values', () => {
    render(
      <VolumeControls
        isMuted
      />
    );

    expect(screen.getByTestId<HTMLDivElement>('volume-controls')).toMatchSnapshot();
  });

  test('snapshot with all the passed props', () => {
    render(
      <VolumeControls isMuted />
    );

    expect(screen.getByTestId<HTMLDivElement>('volume-controls')).toMatchSnapshot();
  });
});
