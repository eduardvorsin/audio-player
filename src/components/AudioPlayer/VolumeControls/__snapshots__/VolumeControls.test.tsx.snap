// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Volume Controls component tests snapshot with all the passed props 1`] = `
<div
  class="sc-gswNZR buozcK"
  data-testid="volume-controls"
>
  <button
    class="sc-bcXHqe eFclmf"
    data-testid="button"
  >
    <svg />
    unmute
  </button>
  <div
    class="sc-dkrFOg kLzxy"
    data-testid="range-slider"
  >
    <input
      style="--progress-percent: 50%;"
      type="range"
    />
  </div>
</div>
`;

exports[`Volume Controls component tests snapshot with default prop values 1`] = `
<div
  class="sc-gswNZR buozcK"
  data-testid="volume-controls"
>
  <button
    class="sc-bcXHqe eFclmf"
    data-testid="button"
  >
    <svg />
    mute
  </button>
  <div
    class="sc-dkrFOg kLzxy"
    data-testid="range-slider"
  >
    <input
      style="--progress-percent: 50%;"
      type="range"
    />
  </div>
</div>
`;
